### PR TITLE
For discussion: Replatform FhirClient on HttpClient

### PR DESCRIPTION
**Pull request by BradBarnich**
I don't have much confidence in these changes yet. I can't get the Integration tests to pass with any of the listed DSTU2 endpoints, even before any changes.

The extent of the public API changes are:
1. the removal of the `Last...` properties from `IFhirClient`, I felt these were problematic because they make `FhirClient` not threadsafe. For test code, it was trivial to write a custom httpclienthandler that captures these.
1. the argument to the event handlers is now a `HttpRequest/ResponseMessage`

Most of the code was back ported from #507 that was merged into develop-r4.

Also addresses #478 

Would such API changes be acceptable? Or must an HttpClient implementation remain separate?

### CODE REVIEW

File: src/Hl7.Fhir.Core/Rest/IFhirClient.cs [hunk 1]
```diff
@@ -13,12 +13,6 @@ public interface IFhirClient
         bool CompressRequestBody { get; set; }
 #endif
         Uri Endpoint { get; }
-        byte[] LastBody { get; }
-        Resource LastBodyAsResource { get; }
-        string LastBodyAsText { get; }
-        HttpWebRequest LastRequest { get; }
-        HttpWebResponse LastResponse { get; }
-        Bundle.ResponseComponent LastResult { get; }
```
Comment (BradBarnich) [inline src/Hl7.Fhir.Core/Rest/IFhirClient.cs]: breaking change

File: src/Hl7.Fhir.Core/Rest/FhirClient.cs [hunk 2]
```diff
@@ -1097,30 +1115,48 @@ private void verifyServerVersion()
...
+        public AfterResponseEventArgs(HttpResponseMessage response)
         {
-            this.RawResponse = webResponse;
-            this.Body = body;
+            Response = response;
         }
 
-        public HttpWebResponse RawResponse { get; internal set; }
-        public byte[] Body { get; internal set; }
+        public HttpResponseMessage Response { get; internal set; }
```
Comment (BradBarnich) [inline src/Hl7.Fhir.Core/Rest/FhirClient.cs]: breaking change

**Comment by brianpos** [comment]
I haven't had a good look over these tests, but one if the things I'd like to see with the client is a pure streaming option that doesn't read the lot into a string first... but that has issues with the existing event handler approach.

**Comment by BradBarnich** [comment]
This doesn't read the body for the event handlers, it could be a stream. But the first step of parsing reads it into a string because that is what the parsers accept, I didn't change that.

It should be possible to read the the content as a stream, create an JsonTextReader/XmlTextReader and pass that to the parser. I can give it a try

**Comment by brianpos** [comment]
There is a streaming version if the parsers too, in fact I believe the string version uses a string stream...

**Comment by brianpos** [comment]
The old code reads it so that the event handler doesn't mess with the stream as it can't rewind...

**Comment by BradBarnich** [comment]
@brianpos I was playing around and I think it should be possible to serialize through a stream directly into the network request. That might reduce allocations a bit, I think one less copy of the xml/json bytes in memory. Did you have a particular use case in mind for a streaming API?

**Pull request closed without merging**